### ISSUE TITLE
Still unable to build demos

### ISSUE THREAD

**Issue by reedkotler** [google/ml-compiler-opt#244]
The instructions are incomplete.

Has anyone built from scratch on a new machine with these instructions?

In an earlier step: export FUCHSIA_SRCDIR=~/fuchsia
~/fuschia will not exist per the instructions.

you need to add:
curl -s "https://fuchsia.googlesource.com/fuchsia/+/HEAD/scripts/bootstrap?format=TEXT" | base64 --decode | bash

Later:
pipenv sync --system
| grep Location | cut -d ' ' -f 2)

This is clearly not right

**Comment by mtrofin** [comment]
We're always welcoming patches, including fixes to the demo script, which definitely could be more precise. We were hoping to eventually encode it as a dockerfile, which should avoid some these issues, too.

**Comment by mtrofin** [comment]
Hmm... so the `pipenv sync` error seems to be the problem. It may have been due to some dependencies' versions being superseded - #218 should hopefully address that. I'm assuming you're using one of the 3 python versions we use in the CI - 3.8, 3.9 or 3.10.

The second part, looks like something may not have been copied correctly. It should be this:

```
TF_PIP=$(python3 -m pip show tensorflow | grep Location | cut -d ' ' -f 2)
```

It should get the path to your python's `site-packages` directory, usually something like `~/.local/python3.10/lib/site-packages`.

**Comment by reedkotler** [comment]
The Readme for the inline demo shows:

cd ~/ml-compiler-opt
sudo apt-get install python3-pip
pip3 install pipenv
pipenv sync --system
| grep Location | cut -d ' ' -f 2)
TF_PIP=$(python3 -m pip show tensorflow 

export TENSORFLOW_AOT_PATH="${TF_PIP}/tensorflow"

export TFLITE_PATH=~/tflite
mkdir ${TFLITE_PATH}
cd ${TFLITE_PATH}
~/ml-compiler-opt/buildbot/build_tflite.sh

**Comment by mtrofin** [comment]
Something's odd - are we talking about this: https://github.com/google/ml-compiler-opt/blob/main/docs/inlining-demo/demo.md#tensorflow-dependencies

This renders like this:

![image](https://github.com/google/ml-compiler-opt/assets/3815402/cac4afa0-0920-4c1b-bdbb-e5f959d16216)

Kind of at a loss - not sure why you have `pipenv sync --system` followed by `| grep` etc. Both on Chrome 113.0.5672.126 and Safari Version 16.4 (18615.1.26.110.1) it renders as in the picture.

What browser are you using?

**Comment by reedkotler** [comment]
i'm using chrome.

ALso:

TF_PIP=$(python3 -m pip show tensorflow | grep Location | cut -d ' ' -f 2)
WARNING: Package(s) not found: tensorflow

**Comment by reedkotler** [comment]
On github i see what you see but not in my local one. Hmmm. I'm using a special environment we have for GPUs in our cloud so that may be doing something wrong here.

**Comment by reedkotler** [comment]
An error occurred while installing zipp==3.15.0 ; python_version >= '3.7' --hash=sha256:112929ad649da941c23de50f356a2b5570c954b65150642bccdd66bf194d224b --hash=sha256:48904fc76a60e542af151aded95726c1a5c34ed43ab4134b597665c86d7ad556! Will try again.
Installing initially failed dependencies.

while executing:

pipenv sync --system

**Comment by reedkotler** [comment]
I think the python version is wrong.

mlxlabz28yb4q7644ebc4a-20230430190655-b6gm58-master:ml-compiler-opt# python --version
Python 2.7.16
mlxlabz28yb4q7644ebc4a-20230430190655-b6gm58-master:ml-compiler-opt# python3 --version
Python 3.7.3

I followed all the instructions. THere was a comment I see that I missed for python versions but the instructions should check for that or explicitly install that.

**Comment by reedkotler** [comment]
I see that python 3.8++ on debian requires building your own from source.

**Comment by mtrofin** [comment]
ack - 3.8, .9, or .10 at the moment.

**Comment by mtrofin** [comment]
Maybe via [bullseye-backports](https://packages.debian.org/bullseye-backports/python/) ?

**Comment by reedkotler** [comment]
I'm doing it from source but there is a way to apt-get it too.  https://linuxconcept.com/how-to-install-python-3-on-debian-10-linux/

**Comment by reedkotler** [comment]
mlxlabz28yb4q7644ebc4a-20230430190655-b6gm58-master:build# ninja distribution
ninja: error: unknown target 'distribution'
mlxlabz28yb4q7644ebc4a-20230430190655-b6gm58-master:build# DESTDIR=${LLVM_INSTALLDIR} ninja install-distribution-stripped
ninja: error: unknown target 'install-distribution-stripped'
mlxlabz28yb4q7644ebc4a-20230430190655-b6gm58-master:build# which ninja
/usr/local/bin/ninja
mlxlabz28yb4q7644ebc4a-20230430190655-b6gm58-master:build# pwd
/root/llvm-project/build

**Comment by mtrofin** [comment]
you may need to nuke the build dir if you ran cmake to it already, and re-run:

```
cmake -G Ninja \
  -DLLVM_ENABLE_LTO=OFF \
  -DLINUX_x86_64-unknown-linux-gnu_SYSROOT=${SYSROOT_DIR}/linux-x64 \
  -DLINUX_aarch64-unknown-linux-gnu_SYSROOT=${SYSROOT_DIR}/linux-arm64 \
  -DFUCHSIA_SDK=${IDK_DIR} \
  -DCMAKE_INSTALL_PREFIX= \
  -DCMAKE_INSTALL_RPATH_USE_LINK_PATH=On \
  -C ${LLVM_SRCDIR}/clang/cmake/caches/Fuchsia-stage2.cmake \
  -C ${TFLITE_PATH}/tflite.cmake \
  ${LLVM_SRCDIR}/llvm
```

**Comment by reedkotler** [comment]
It was clean but I nuked it and rebuilt with the same result

-- Build files have been written to: /root/llvm-project/build
mlxlabz28yb4q7644ebc4a-20230430190655-b6gm58-master:build# ninja distribution
ninja: error: unknown target 'distribution'

**Comment by reedkotler** [comment]
-- Configuring done
-- Generating done
-- Build files have been written to: /root/llvm-project/build

This was the result of the Cmake

**Comment by reedkotler** [comment]
Is anyone actually running these demos? I have just been trying the inliner.

**Comment by mtrofin** [comment]
Afaik no one is running them regularly, if that's what you mean. The demos are maintained as a "best effort" thing. Their main goal is to point at how to build clang with ml support enabled, collect a corpus, run the training loop, and embed the resulting model. 

Looking into what may cause your build to fail.

**Comment by reedkotler** [comment]
My manager wants me to do some kind of demo with MLGO. It's not so important what it is at this time.
I'm going to walk them through the paper a bit.

**Comment by mtrofin** [comment]
For a presentation-like thing, the demos here would be unsuitable, because the training part may well take ~a day :)

**Comment by reedkotler** [comment]
well, i can show the steps and skip the actual development because i have the results already.

**Comment by reedkotler** [comment]
maybe all i can really do is a talk then

**Comment by reedkotler** [comment]
can i run LLVM with prebuilt inliner or register allocator? and compare that against another?

**Comment by mtrofin** [comment]
You'd need to build it with those models, and mileage may vary - i.e. the models we built were trained on source that could be sufficiently different from what you have.

To build embedding the inlining Oz model, for example:
- download the model from https://github.com/google/ml-compiler-opt/releases/tag/inlining-Oz-v1.1, unzip it somewhere. Let's call the path to the top directory that contains the `saved_model.pb` file `$MODEL`
- install tensorflow (`pip install --user tensorflow`)
- add `-DTENSORFLOW_AOT_PATH=$HOME/.local/lib/python3.10/site-packages/tensorflow -DLLVM_INLINER_MODEL_PATH=$MODEL` (assuming python 3.10) to your cmake
- build clang.
 
pass `-mllvm -enable-ml-inliner=release` to opt into using the model. The model assumes you're building `-Os` or `-Oz`.

**Comment by reedkotler** [comment]
So the idea is that you would normally train it on source for the project you want to use it for?

**Comment by mtrofin** [comment]
Yes, and it doesn't have to be exactly that source, nor does it need retaining as you work on the project, nor as the compiler changes. It's more about feature values distributions, meaning that as long as the code looks similar enough to what you trained on, you should see size reduction. We are still using the model I linked (almost 2 years old) for internal binaries where size matters, for example. 

What "similar enough" means, not clear yet. Interesting topic for ML research, I suppose.

**Comment by Sharanya-Saha** [comment]
Hi @reedkotler, I have been trying to build a MLGO based clang from a long time now. I have tried training it using the demo instructions. But I don't see the results of using my trained model outside Fuchsia. Does the pre-trained model do something general?

**Comment by reedkotler** [comment]
Unfortunately, I've left ByteDance and don't have my notes now but I think I can answer this but i need some time to refresh my memory on this.

You don't want to try to recreate their work from Fuchsia. You will never get that all to work.

You can use the Google pre-trained models, or you can train your own model. I don't see any point in the beginning of training your own models for this.

Do you want to use the register optimizer or the inline optimizer for your clang?

I'm very busy at the moment with something else, but I will try and go back through my online posts, which I believe had all that I knew in it.




________________________________

**Comment by Sharanya-Saha** [comment]
I wanted to try it for inlining for size. I followed the instructions available online and got a trained model using Fuchsia. I don't see much results on using the release build made from the one trained using Fuchsia(Mentioned in Demo). How do I use the pretrained model? Are the commands available anywhere? I have tried to use similar instructions as I did for the one with fuchsia, but that has some dependency with the fuchsia cmake, it seems.

**Comment by mtrofin** [comment]
You can use the pretrained model after you download it from [here](https://github.com/google/ml-compiler-opt/releases/tag/inlining-Oz-v1.1), by substituting the model you would have trained in the demo with it, [here](https://github.com/google/ml-compiler-opt/blob/main/docs/inlining-demo/demo.md#deploying-and-using-the-new-policy).

Note the pretrained model was trained on x86_64 server side code, meaning that's the kind of code where it would do best (albeit "server side code" is quite ambiguous - specifically, feature value distributions and how they get quantized - the "vocab" step - is what matters to (roughly) match)

For the model you trained, how did the `reward_distribution` look like when you stopped the training?

**Comment by boomanaiden154** [comment]
We are very willing to review PRs patching up the demos if you're interested in sending them.
